### PR TITLE
DE TN Fixes

### PR DESCRIPTION
**Pull request by zoobereq**
# What does this PR do ?

This PR implements DE TN fixes for the following issues:

- Adds support for normalizing social media tags (e.g. `@zoobereq` and `@zoobereq.net`)
- Adds support for normalizing comma-separated digit strings
- Adds support for period-separated time formats (e.g. `2.30` and `02.30`)
- Fixes the issue where the final period in sentences ending with a domain name would normalize as part of that domain.

This PR *does not* address the following:

- [The issue with optionally period-separated cardinals](https://github.com/NVIDIA/NeMo-text-processing/issues/171) (e.g. 1 Mil = 100000 | 1.000.000) , which the currently implemented DE TN/ITN doesn't support. The TN and ITN taggers have been rebuilt to accommodate for that hiccup, but since cardinals plug into almost all other DE classes, this issue will be addressed gradually as other classes are fixed/(re)developed.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Have you signed your commits? Use ``git commit -s`` to sign.
- [x] Do all unittests finish successfully before sending PR?
   1) ``pytest`` or (if your machine does not have GPU) ``pytest --cpu`` from the root folder (given you marked your test cases accordingly `@pytest.mark.run_only_on('CPU')`).
   2) Sparrowhawk tests ``bash tools/text_processing_deployment/export_grammars.sh --MODE=test ...``
- [x] If you are adding a new feature: Have you added test cases for both `pytest` and Sparrowhawk [here](tests/nemo_text_processing).
- [x] Have you added ``__init__.py`` for every folder and subfolder, including `data` folder which has .TSV files?
- [x] Have you followed codeQL results and removed unused variables and imports (report is at the bottom of the PR in github review box) ?
- [x] Have you added the correct license header `Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.` to all newly added Python files?
- [x] If you copied [nemo_text_processing/text_normalization/en/graph_utils.py](nemo_text_processing/text_normalization/en/graph_utils.py) your header's second line should be `Copyright 2015 and onwards Google, Inc.`. See an example [here](https://github.com/NVIDIA/NeMo-text-processing/blob/main/nemo_text_processing/text_normalization/en/graph_utils.py#L2).
- [x] Remove import guards (`try import: ... except: ...`) if not already done.
- [x] If you added a new language or a new feature please update the [NeMo documentation](https://github.com/NVIDIA/NeMo/blob/main/docs/source/nlp/text_normalization/wfst/wfst_text_normalization.rst) (lives in different repo).
- [x] Have you added your language support to [tools/text_processing_deployment/pynini_export.py](tools/text_processing_deployment/pynini_export.py).
  

**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation
- [x] Test

If you haven't finished some of the above items you can still open "Draft" PR.

### CODE REVIEW

**Comment by ekmb** [comment]
@zoobereq please update grammars path in Jenkins to re-built CI grammars https://github.com/NVIDIA/NeMo-text-processing/blob/main/Jenkinsfile#L15

File: nemo_text_processing/text_normalization/de/taggers/electronic.py [hunk 1]
```diff
@@ -35,31 +35,44 @@ def __init__(self, deterministic: bool = True):
...
-        accepted_common_domains = pynini.union(*accepted_common_domains)
-        accepted_symbols = [x[0] for x in load_labels(get_abs_path("data/electronic/symbols.tsv"))]
-        accepted_symbols = pynini.union(*accepted_symbols) - dot
-        accepted_characters = pynini.closure(NEMO_ALPHA | NEMO_DIGIT | accepted_symbols)
+
+        all_symbols = [x[0] for x in load_labels(get_abs_path("data/electronic/symbols.tsv"))]
+        all_symbols = pynini.union(*all_symbols)
+        # all symbols
+        accepted_symbols = pynini.difference(all_symbols, dot)  # alphabet of accepted symbols excluding the '.'
+        accepted_characters = pynini.closure(
```
Comment (bonham79) [inline nemo_text_processing/text_normalization/de/taggers/electronic.py:43]: Need more specific variable names. `all` vs `accepted` is awkward.

File: nemo_text_processing/text_normalization/de/taggers/time.py [hunk 2]
```diff
@@ -43,37 +43,35 @@ def __init__(self, deterministic: bool = True):
         labels_minute_single = [str(x) for x in range(1, 10)]
         labels_minute_double = [str(x) for x in range(10, 60)]
 
-        delete_leading_zero_to_double_digit = (pynutil.delete("0") | (NEMO_DIGIT - "0")) + NEMO_DIGIT
+        delete_leading_zero_to_double_digit = (pynutil.delete("0").ques | (NEMO_DIGIT - "0")) + NEMO_DIGIT
```
Comment (bonham79) [inline nemo_text_processing/text_normalization/de/taggers/time.py:46]: `.ques` is unnecessary ( and potentially damaging here).

File: nemo_text_processing/text_normalization/de/taggers/tokenize_and_classify.py [hunk 3]
```diff
@@ -132,7 +132,7 @@ def __init__(
                 pynini.closure(punct + pynutil.insert(" ")) + token + pynini.closure(pynutil.insert(" ") + punct)
             )
 
-            graph = token_plus_punct + pynini.closure(pynutil.add_weight(delete_extra_space, 1.1) + token_plus_punct)
+            graph = token_plus_punct + pynini.closure((delete_extra_space).ques + token_plus_punct)
```
Comment (bonham79) [inline nemo_text_processing/text_normalization/de/taggers/tokenize_and_classify.py:135]: Did you stress test to make sure the lack of a weight is fine?

File: nemo_text_processing/text_normalization/de/verbalizers/electronic.py [hunk 4]
```diff
@@ -55,21 +55,21 @@ def add_space_after_char():
...
         protocol = (
-            pynutil.delete("protocol: \"")
+            pynutil.delete('protocol: "')
             + add_space_after_char() @ pynini.cdrewrite(graph_symbols, "", "", NEMO_SIGMA)
-            + pynutil.delete("\"")
+            + pynutil.delete('"')
         )
         self.graph = (pynini.closure(protocol + pynini.accep(" "), 0, 1) + domain) | (
-            user_name + pynini.accep(" ") + pynutil.insert("at ") + domain
+            user_name + pynini.accep(" ") + pynutil.insert("at ") + domain | (pynutil.insert("at ") + user_name)
```
Comment (bonham79) [inline nemo_text_processing/text_normalization/de/verbalizers/electronic.py:72]: Use NeMo Space here for readability

File: tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt [hunk 5]
```diff
@@ -6,4 +6,11 @@ a eins b zwei at a b c punkt com~a1b2@abc.com
...
 h t t p s doppelpunkt slash slash w w w punkt a b c punkt com~https://www.abc.com
 w w w punkt a b c punkt com~www.abc.com
-h t t p s doppelpunkt slash slash w w w punkt a b c punkt com slash a b fragezeichen gleichheitszeichen drei bindestrich slash a b s slash eins~https://www.abc.com/ab?=3-/abs/1
\ No newline at end of file
+b r e t t s p i e l v e r s a n d punkt de .~brettspielversand.de.
+w w w punkt z u b e r e k punkt net .~www.zuberek.net.
+w w w punkt a m a z o n punkt com punkt de .~www.amazon.com.de.
+h t t p s doppelpunkt slash slash w w w punkt a b c punkt com slash a b fragezeichen gleichheitszeichen drei bindestrich slash a b s slash eins~https://www.abc.com/ab?=3-/abs/1
+at z u c k~@zuck
+at z o o b e r e q~@zoobereq
```
Comment (bonham79) [inline tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt:14]: Don't use your name as an example. It leads to potential doxing.

File: tests/nemo_text_processing/de/test_sparrowhawk_inverse_text_normalization.sh [hunk 6]
```diff
@@ -3,7 +3,7 @@
 PROJECT_DIR=/workspace/tests
 
 GRAMMARS_DIR=${1:-"/workspace/sparrowhawk/documentation/grammars"}
-PROJECT_DIR=${2:-"/workspace/tests/en"}
+PROJECT_DIR=${2:-"/workspace/tests/"}
```
Comment (bonham79) [inline tests/nemo_text_processing/de/test_sparrowhawk_inverse_text_normalization.sh:6]: Why is this necessary?

File: tests/nemo_text_processing/de/test_sparrowhawk_normalization.sh [hunk 7]
```diff
@@ -1,7 +1,7 @@
 #! /bin/sh
 
 GRAMMARS_DIR=${1:-"/workspace/sparrowhawk/documentation/grammars"}
-PROJECT_DIR=${2:-"/workspace/tests/en"}
+PROJECT_DIR=${2:-"/workspace/tests/"}
```
Comment (bonham79) [inline tests/nemo_text_processing/de/test_sparrowhawk_normalization.sh:4]: Ditto

**Review by bonham79 — CHANGES_REQUESTED** [review]

File: nemo_text_processing/text_normalization/de/taggers/electronic.py (hunk 1 above)
Comment (zoobereq) [reply nemo_text_processing/text_normalization/de/taggers/electronic.py:43]: Will fix.

File: nemo_text_processing/text_normalization/de/taggers/time.py (hunk 2 above)
Comment (zoobereq) [reply nemo_text_processing/text_normalization/de/taggers/time.py:46]: Happy to remove it, but without it, comma-separated digit strings (e.g. `1,2,3,4`) won't normalize , causing the tests to fail.

File: nemo_text_processing/text_normalization/de/taggers/tokenize_and_classify.py (hunk 3 above)
Comment (zoobereq) [reply nemo_text_processing/text_normalization/de/taggers/tokenize_and_classify.py:135]: Yes.  It seemed to work fine with whatever I threw at it.

File: nemo_text_processing/text_normalization/de/verbalizers/electronic.py (hunk 4 above)
Comment (zoobereq) [reply nemo_text_processing/text_normalization/de/verbalizers/electronic.py:72]: Okey dokey

File: tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt (hunk 5 above)
Comment (zoobereq) [reply tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt:14]: Good point - thank you.  Will fix.

File: tests/nemo_text_processing/de/test_sparrowhawk_inverse_text_normalization.sh (hunk 6 above)
Comment (zoobereq) [reply tests/nemo_text_processing/de/test_sparrowhawk_inverse_text_normalization.sh:6]: Sparrowhawk throws an error otherwise.  I described it [here](https://docs.google.com/document/d/1IWm_rLK0dHe0E2aewdAf63EW3doV47JLV52e_yRsyLs/edit).

File: tests/nemo_text_processing/de/test_sparrowhawk_normalization.sh (hunk 7 above)
Comment (zoobereq) [reply tests/nemo_text_processing/de/test_sparrowhawk_normalization.sh:4]: Right, same thing.  Once I removed it, I was able to run Sparrowhawk tests just fine.

File: nemo_text_processing/text_normalization/de/taggers/electronic.py (hunk 1 above)
Comment (zoobereq) [reply nemo_text_processing/text_normalization/de/taggers/electronic.py:43]: Fixed

File: nemo_text_processing/text_normalization/de/taggers/time.py (hunk 2 above)
Comment (zoobereq) [reply nemo_text_processing/text_normalization/de/taggers/time.py:46]: Without applying a closure over at most one occurrence time expressions in the following format `h.mm` will not normalize correctly.

File: nemo_text_processing/text_normalization/de/taggers/electronic.py (hunk 1 above)
Comment (bonham79) [reply nemo_text_processing/text_normalization/de/taggers/electronic.py:43]: Did you push? Not showing up

File: nemo_text_processing/text_normalization/de/taggers/time.py (hunk 2 above)
Comment (tbartley94) [reply nemo_text_processing/text_normalization/de/taggers/time.py:46]: Oh this is for time, nvm. Confusing with cardinal.

File: tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt [hunk 8]
```diff
@@ -6,4 +6,10 @@ a eins b zwei at a b c punkt com~a1b2@abc.com
 a b drei bindestrich s d d bindestrich drei at g mail punkt com~ab3-sdd-3@gmail.com
 h t t p s doppelpunkt slash slash w w w punkt a b c punkt com~https://www.abc.com
 w w w punkt a b c punkt com~www.abc.com
-h t t p s doppelpunkt slash slash w w w punkt a b c punkt com slash a b fragezeichen gleichheitszeichen drei bindestrich slash a b s slash eins~https://www.abc.com/ab?=3-/abs/1
\ No newline at end of file
+b r e t t s p i e l v e r s a n d punkt de .~brettspielversand.de.
+w w w punkt z u b e r e k punkt net .~www.zuberek.net.
```
Comment (tbartley94) [inline tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt:10]: remove your name

File: tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt [hunk 9]
```diff
@@ -6,4 +6,10 @@ a eins b zwei at a b c punkt com~a1b2@abc.com
...
+w w w punkt a m a z o n punkt com punkt de .~www.amazon.com.de.
+h t t p s doppelpunkt slash slash w w w punkt a b c punkt com slash a b fragezeichen gleichheitszeichen drei bindestrich slash a b s slash eins~https://www.abc.com/ab?=3-/abs/1
+at z u c k~@zuck
```
Comment (tbartley94) [inline tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt:13]: use @jensen

**Review by tbartley94 — APPROVED** [review]
Remove references to yourself and competitors and LGTM

File: tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt (hunk 8 above)
Comment (zoobereq) [reply tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt]: Done.

File: tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt (hunk 9 above)
Comment (zoobereq) [reply tests/nemo_text_processing/de/data_text_normalization/test_cases_electronic.txt]: Done.

File: tests/nemo_text_processing/de/test_sparrowhawk_inverse_text_normalization.sh (hunk 6 above)
Comment (ekmb) [reply tests/nemo_text_processing/de/test_sparrowhawk_inverse_text_normalization.sh:6]: @zoobereq could you please add DE tests to SH? I don't see them in Jenkins
Have you tried `/workspace/tests/de`?

Comment (zoobereq) [reply tests/nemo_text_processing/de/test_sparrowhawk_inverse_text_normalization.sh:6]: @ekmb they are referenced in `de/test_sparrowhawk_normalization.sh` and `de/test_sparrowhawk_inverse_text_normalization.sh`.  They are also referenced in `JenkinsFile` -- what am I missing?
And yes, I did try `/workspace/tests/de` and received:
`test_sparrowhawk_normalization.sh: line 86: /workspace/tests/en/../shunit2/shunit2: No such file or directory
` (also described [here](https://docs.google.com/document/d/1IWm_rLK0dHe0E2aewdAf63EW3doV47JLV52e_yRsyLs/edit))

**Comment by tbartley94** [comment]
@zoobereq LGTM, will approve once Evelina is happy

**Review by ekmb — APPROVED** [review]
Thanks!
